### PR TITLE
Adds LoRA support including cache warmup

### PR DESCRIPTION
**Pull request by dans-msft**
This change adds support for LoRA finetuning to llgtrt.

This implementation of LoRA support stays close to the support already provided by TensorRT.  It can be activated by adding two parameters to your inference requests:

_lora_id_: unsigned integer identifier for the set of LoRA weights.  If the set of weights is already present in this TensorRT instance's cache then providing this Id is enough.  If not then you must also specify _lora_dir_.

_lora_dir_: name of the LoRA subdirectory that contains the weights.  You must also provide _lora_dir_ when using this parameter.  Use of _lora_dir_ has the side effect of loading this set of weights into this TensorRT instance's cache, keyed by the provided _lora_id_.  If the cache fills up then TensorRT may evict another set of weights.

### CODE REVIEW

File: llgtrt/src/routes/completions.rs [hunk 1]
```diff
@@ -123,6 +126,74 @@ fn req_params_from_openai(params: &CommonCreateParams) -> Result<RequestParams>
...
+    }
+    
+    // Reshape the array to the new squeezed shape
+    array.into_shape_with_order(IxDyn(&squeezed_shape)).expect("Failed to reshape array")
+}
+
+fn req_lora_params_from_openai(params: &CommonCreateParams) -> Result<Option<LoraParams>> {
+    if let Some(lora_id) = params.lora_id {
+        if let Some(lora_dir) = &params.lora_dir {
+            let base_path: &Path = Path::new("/lora");
```
Comment (mmoskal) [inline llgtrt/src/routes/completions.rs]: ideally, we would pass "/lora" as command line arg (which could default to "/lora")

File: llgtrt/src/routes/completions.rs [hunk 2]
```diff
@@ -123,6 +126,74 @@ fn req_params_from_openai(params: &CommonCreateParams) -> Result<RequestParams>
...
+            let lora_path = base_path.join(lora_dir);
```
Comment (mmoskal) [inline llgtrt/src/routes/completions.rs]: I would make sure users don't pass '../../../etc/passwd' or whatever

```rust
    let lora_path = lora_path.canonicalize()?;
    if !lora_path.starts_with(base_path) {
        return Err(anyhow::anyhow!("lora_dir is outside the allowed /lora directory"));
    }
```

File: LoRA.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+### Converting LoRA weights
+
+TensorRT cannot read LoRA weights directly from Huggingface safetensors format. So we convert weights to a simplified "tensor" format that TensorRT can read.  Note that these weights are simply converted -- unlike the base model they are not compiled into an execution graph.  TensorRT provides the _hf_lora_convert.py_ utility to perform this conversion.  This step should be done within the llgtrt container.
+
+```bash
+# Create a directory for LoRA weights
+mkdir -p /models/lora/my_finetuning
+
+# Convert weights.  Float32 is used only as an intemediate storage type here; the weights will be converted to bfloat16 when loaded for inference.
```
Comment (mmoskal) [inline LoRA.md]: why use f32 as storage format? wouldn't bf16 be 50% smaller?

File: llgtrt/src/routes/completions.rs (hunk 1 above)
Comment (mmoskal) [reply llgtrt/src/routes/completions.rs]: or actually best if it doesn't default to anything, so lora is disabled by default

File: llgtrt/src/routes/completions.rs [hunk 4]
```diff
@@ -123,6 +128,78 @@ fn req_params_from_openai(params: &CommonCreateParams) -> Result<RequestParams>
...
+                    lora_id: lora_cache.resolve_id(lora_model),
+                    weights: None,
+                    config: None,
+                }))
+            }
+        } else {
+            Ok(None)
+        }
+    } else {
+        Ok(None)
```
Comment (mmoskal) [inline llgtrt/src/routes/completions.rs]: might be good to return Err() if lora_root is none and lora_model is not

File: llgtrt/src/routes/completions.rs [hunk 5]
```diff
@@ -223,27 +373,11 @@ async fn mk_req_info(
             llg.log_json_progress = true;
         }
 
-        // temperature handled by logits processing - this has to be 1.0
-        // to avoid double-application of temperature
+        // temperature handled by logits processing
```
Comment (mmoskal) [inline llgtrt/src/routes/completions.rs]: would be good to keep these three comments

File: llgtrt/src/routes/completions.rs (hunk 1 above)
Comment (dans-msft) [reply llgtrt/src/routes/completions.rs]: Done (both suggestions)

File: LoRA.md (hunk 3 above)
Comment (dans-msft) [reply LoRA.md]: Done by switching to safetensors.  bfloat16 is fully supported!

File: llgtrt/src/routes/completions.rs (hunk 5 above)
Comment (dans-msft) [reply llgtrt/src/routes/completions.rs]: Not sure how that got removed but ... done.

File: llgtrt/src/routes/completions.rs (hunk 4 above)
Comment (dans-msft) [reply llgtrt/src/routes/completions.rs]: Done

File: llgtrt/src/routes/completions.rs (hunk 2 above)
Comment (dans-msft) [reply llgtrt/src/routes/completions.rs]: I added a step where I canonicalize the path and then ensure it is under the base /lora directory as you suggested.  Note that canonicalize will fail if the path doesn't exist so I added better error handling there too.

File: LoRA.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
+# LoRA support in llgtrt
+
+TensorRT provides full support for parameter-efficient finetuning via a technicque called _Low Rank Adaptation_, commonly abbreviated as _LoRA_.  The engine even provides support for multiple simultaneous sets of finetuned weights (often known as _Multi-LoRA_), though configuring this requires a few extra steps, both during setup and inference.
+
+The following extra steps are required during setup:
+
+1. Compile the base model weights with the LoRA module activated.
```
Comment (mmoskal) [inline LoRA.md]:
```suggestion
1. Build the base model engine with the LoRA module activated.
```

(using trtllm terminology)

File: LoRA.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+2. Convert each set of LoRA weights from standard Huggingface format to a specialized tensor format compatible with TensorRT.  Store these in a directory accessible from your TensorRT installation.
+3. Launch llgtrt/TensorRT with an extra parameter pointing to your LoRA weights.
+
+During inference, TensorRT maintains a cache of LoRA weights within GPU memory where the precise number of LoRA models that can fit in the cache varies based on your configuration.  To perform inference using a specific LoRA model you must first load its weights into the cache keyed by a unique integer identifier.  Once weights are loaded then subsequent inference requests may be made using the integer Id alone.  Depending on the size of your cache, loading a new set of LoRA weights may result in another set being evicted from the cache.
```
Comment (mmoskal) [inline LoRA.md]: Do we still want the interger ID references here?

File: LoRA.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+cp /models/Meta-Llama-3.1-8B-Instruct/tokenizer.json /models/model-engine
+cp /models/Meta-Llama-3.1-8B-Instruct/tokenizer_config.json /models/model-engine
+
+# Exit the container
+exit
+```
+
+### Extract LoRA weights
+
+TensorRT cannot read LoRA weights directly from Huggingface safetensors format. So we provide a Python-based utility to extract weights into a simplified "tensor" format that TensorRT can read.
```
Comment (mmoskal) [inline LoRA.md]:
```suggestion
TensorRT cannot read LoRA weights directly from Huggingface format. So we provide a Python-based utility to extract weights into a simplified format (also safetensors-based) that TensorRT can read.
```

File: llgtrt/src/routes/completions.rs [hunk 9]
```diff
@@ -228,22 +384,7 @@ async fn mk_req_info(
...
         req_params.temperature = 1.0;
 
-        // Test argmax
-        // llg.temperature = 1.0;
-        // req_params.temperature = 0.0;
-        // req_params.top_k = 1;
-
         req_params.use_logits_post_processor = true;
-
-        // If we do that, we need to make sure we return the tokens forced
```
Comment (mmoskal) [inline llgtrt/src/routes/completions.rs:238]: woud be good to restore this comment

File: trtllm_rs/src/tlc.rs [hunk 10]
```diff
@@ -169,6 +198,128 @@ impl Default for ffi::TlcEngineParams {
...
+const TLC_I8: i32 = 2;
+const TLC_I32: i32 = 3;
+const TLC_I64: i32 = 4;
+const TLC_BF16: i32 = 5;
+//const TLC_F8: i32 = 6;  // TODO: find a way to support F8 correctly through Rust.
+const TLC_F16: i32 = 7;
+const TLC_F32: i32 = 8;
+const TLC_UNKNOWN: i32 = 9;
+
+fn _tlc_get_data_type<T: 'static>() -> i32 {
```
Comment (mmoskal) [inline trtllm_rs/src/tlc.rs]: please don't use leading `_` in rust private function names, as it (I've just learned) prevents "unused function" warnings

File: trtllm_rs/src/tlc.rs [hunk 11]
```diff
@@ -169,6 +198,128 @@ impl Default for ffi::TlcEngineParams {
...
+        Dtype::I32 => TLC_I32,
+        Dtype::I64 => TLC_I64,
+        Dtype::BF16 => TLC_BF16,
+        Dtype::F16 => TLC_F16,
+        Dtype::F32 => TLC_F32,
+        _ => TLC_UNKNOWN,
+    }
+}
+
+fn _tlc_extract_arrayd<T: 'static>(array: ArrayD<T>) -> (ffi::TlcTensor, Vec<i64>, Vec<T>)
```
Comment (mmoskal) [inline trtllm_rs/src/tlc.rs]: this is no longer needed and so is the `ndarray` dependency

File: trtllm_rs/src/tlc.rs [hunk 12]
```diff
@@ -194,18 +345,44 @@ impl Executor {
...
         );
-        let arg = ffi::TlcRequest {
+
+        // These declarations make sure that the shape and data vectors extracted from the tensors live long
+        // enough to survive the call to tlc_enqueue_request.  Otherwise the data we pass through to the C++
+        // layer will become invalid.
+        let _weights_shape: Vec<i64>;
+        let _weights_vec: Vec<bf16>;
+        let _config_shape: Vec<i64>;
+        let _config_vec: Vec<i32>;
```
Comment (mmoskal) [inline trtllm_rs/src/tlc.rs]: are these still used?

**Review by mmoskal — COMMENTED** [review]
Just a few small things left!

File: LoRA.md (hunk 6 above)
Comment (dans-msft) [reply LoRA.md]: Done.  And I also updated the term "Convert" to "Extract" since that's the new name of that utility.

And impressed to see comments on documentation!

File: LoRA.md (hunk 7 above)
Comment (dans-msft) [reply LoRA.md]: Fixed

File: llgtrt/src/routes/completions.rs (hunk 9 above)
Comment (dans-msft) [reply llgtrt/src/routes/completions.rs:238]: Fixed.  Wonder how this got lost?  (stray keystroke into a Visual Studio that stole focus perhaps)

File: trtllm_rs/src/tlc.rs (hunk 10 above)
Comment (dans-msft) [reply trtllm_rs/src/tlc.rs]: Done

File: trtllm_rs/src/tlc.rs (hunk 11 above)
Comment (dans-msft) [reply trtllm_rs/src/tlc.rs]: Done

Comment (dans-msft) [reply trtllm_rs/src/tlc.rs]: Also got rid of half and ndarray from Cargo.toml (both here and in the main llgtrt project)

File: trtllm_rs/src/tlc.rs (hunk 12 above)
Comment (dans-msft) [reply trtllm_rs/src/tlc.rs]: Nope.  They were part of a hacky way to keep some data alive before I switched over to borrowing ReqInfo.

File: LoRA.md (hunk 8 above)
Comment (dans-msft) [reply LoRA.md]: Done

**Review by mmoskal — APPROVED** [review]
Awesome!
